keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2868 @@
+The Kubernetes admission chain also scales better, and is more resiliant, when
+matching is precise and the validation expressions don't need to do any
+post-matching checks that could have been handled by matching.
+
+Note that if a policy binding is to be applied as broadly as possible (i.e.
+everywhere allowed by the policy definition) it must do so by using a wildcard
+match rule.
+
+Match Policy:
+
+`MatchPolicy` will work the same as for admission webhooks. It will default to
+`Equivalent` but may be set to `Exact`. See "Use Case: Multiple policy
+definitions for different versions of CRD" for an explanation of why we need
+`MatchPolicy`.
+
+xref:
+
+- https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules
+- https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/
+- https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#resourcepolicyrule-v1beta1-flowcontrol-apiserver-k8s-io
+- https://github.com/open-policy-agent/gatekeeper/blob/a1add93b0beb5c48eb92a6a2eb5ee7d21551a1b6/pkg/mutation/match/match.go#L22-L56
+
+#### Reporting violations to Clients
+
+<<[UNRESOLVED jpbetz, TristonianJones ]>>
+Use CEL expressions for messages instead of for message args?
+<<[/UNRESOLVED]>>
+
+This section focuses on how information is reported back to clients in
+when validations fail.
+
+Goals:
+
+- Feature parity with `AdmissionReview`
+  - Support allow/deny result, warnings, audit annotations
+  - Support for reasons/codes
+- Ability to format message strings
+
+High level proposal:
+
+- Each policy may set a `denyReason` and/or `denyCode`
+- Each validation may define a message:
+  - `message` - may contain `{1}` format args that are supplied by
+    `messageArgs: [<cel expression>, <cel expression>, ...]`
+    - If `message` is absent, `expression` and `name` will be included in the
+      failure message
+    - If any of the arg CEL expressions fail: `expression` and `name` will be
+      included in the failure message plus the arg evaluation failure
+- allow/deny, warnings, audit annotations will be collectively referred to as
+  `enforcement`, which will have the following options:
+  - `Deny`,
+  - `Warn`
+  - `Audit`
+  - (metrics will be emitted regardless of which enforcement options are set, so
+    `enforcement: []` may be useful)
+- Policy definitions identify validation expressions by name so that if the
+  enforcement is `Audit` the name can be used as the audit annotation key.
+- Policy bindings set enforcement options
+
+Example policy definition:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  ...
+  denyReason: Invalid
+  vars: # MUST BE A REAL LIST WITH ORDERING
+    - containers: "self.containers + self.initContainers"
+    - foo: "container.size()"
+  validations:
+    - expression: "self.name.startsWith('xyz-')"
```

Comment:
Do you short-circuit the validation expressions blocks or aggregate the violations messages together? My guess is that you accumulate the validation violations, but it's probably worth spelling out and whether `denyReason` affects the evaluation behavior. For example a `denyReason: Deny` could potentially short-circuit (and this also depends on whether audit logging is always required on deny).

Reply:
This is a good question, deny/admit and short-circuit/continue seem like orthogonal options and maybe there should be two fields. Maybe more. API types in the KEP are never final, that happens when we do a detailed api review. In fact maybe we should just expect to change this after we do an alpha and see what some real-ish policies actually want to do.

Reply:
Short term we plan to aggregate violations. Long term is up in the air. There are some use cases where short-circuiting seems desirable (failed authz check should probably conceal any other validation errors..). I've made a note in future work.